src/lib/table/table-data-source.ts

```diff
@@ -0,0 +1,209 @@
+  }
+
+  /**
+   * Subscribe to changes that should trigger an update to the table's rendered rows. When the
+   * changes occur, process the current state of the filter, sort, and pagination along with
+   * the provided base data and send it to the table for rendering.
+   */
+  _updateChangeSubscription() {
+    // Sorting and/or pagination should be watched if MdSort and/or MdPaginator are provided.
+    // Otherwise, use an empty observable stream to take their place.
+    const sortChange = this._sort ? this._sort.mdSortChange : observableOf();
+    const pageChange = this._paginator ? this._paginator.page : observableOf();
+
+    if (this._renderChangesSubscription) { this._renderChangesSubscription.unsubscribe(); }
+    this._renderChangesSubscription = RxChain.from(this._data)
+        // Watch for base data or filter changes to provide a filtered set of data.
+        .call(combineLatest, this._filter)
+        .call(map, ([data]) => this._filterData(data.slice()))
+        // Watch for filtered data or sort changes to provide an ordered set of data.
+        .call(combineLatest, startWith.call(sortChange, null))
+        .call(map, ([data]) => this._orderData(data.slice()))
+        // Watch for ordered data or page changes to provide a paged set of data.
+        .call(combineLatest, startWith.call(pageChange, null))
+        .call(map, ([data]) => this._pageData(data.slice()))
+        // Watched for paged data changes and send the result to the table to render.
+        .subscribe(data => this._renderData.next(data.slice()));
+  }
+
+  /**
+   * Returns a filtered data array where each filter object contains the filter string within
+   * the result of the filterTermAccessor function.
+   */
+  _filterData(data: T[]) {
+    // If there is a filter string, filter out data that does not contain it.
+    // Each data object is converted to a string using the function defined by filterTermAccessor.
+    // May be overriden for customization.
+    const filteredData = this.filter === '' ? data : data.filter(obj => {
+      return this.filterTermAccessor(obj).indexOf(this.filter) != -1;
+    });
+
+    if (this.paginator) { this._updatePaginator(filteredData.length); }
+
+    return filteredData;
+  }
+
+  /**
+   * Sorts the provided array of data, unless an instance of MdSort has not been provided
+   * or if it does not have an active key or direction.
+   * Uses the default data accessor for data lookup, unless a sortDataAccessor function is defined.
+   */
+  _orderData(data: T[]): T[] {
+    // If there is no active sort or direction, return the data without trying to sort.
+    if (!this.sort || !this.sort.active || this.sort.direction == '') { return data; }
+
+    return data.sort((a, b) => {
+      let valueA = this.sortingDataAccessor(a, this.sort!.active);
+      let valueB = this.sortingDataAccessor(b, this.sort!.active);
+      return (valueA < valueB ? -1 : 1) * (this.sort!.direction == 'asc' ? 1 : -1);
+    });
```

Comment:
You could avoid the non-nullability assertion operator by storing a reference to `this.sort` 
```ts
if (!this.sort || !this.sort.active || this.sort.direction == '') { 
  return data; 
}

const sort = this.sort;
return data.sort((a, b) => {
  let valueA = this.sortingDataAccessor(a, sort.active);
  let valueB = this.sortingDataAccessor(b, sort.active);
  return (valueA < valueB ? -1 : 1) * (sort.direction == 'asc' ? 1 : -1);
});
```